File: src/platforms/host.ts
import { useEffect, useState } from "react";
import { invoke } from "@tauri-apps/api/core";

/** The machine the app is running on. */
export type Host = {
  /** `linux`, `macos`, `windows` or `android`. */
  os: string;
  desktop: boolean;
};

// cannot change while the app is open
let cached: Host | null = null;
let request: Promise<void> | null = null;
const watchers = new Set<() => void>();

function announce() {
  for (const watcher of watchers) watcher();
}

function load(): Promise<void> {
  request ??= invoke<Host>("host_info")
    .then((host) => {
      cached = host;
    })
    .catch(() => {
      // a failure should leave the desktop settings in place, not hide them
      cached = { os: "unknown", desktop: true };
    })
    .finally(announce);
  return request;
}

void load();

/** The host, or null until the first answer arrives. */
export function useHost(): Host | null {
  const [host, setHost] = useState<Host | null>(cached);

  useEffect(() => {
    if (cached) {
      setHost(cached);
      return;
    }
    const read = () => setHost(cached);
    watchers.add(read);
    void load();
    return () => {
      watchers.delete(read);
    };
  }, []);

  return host;
}

/**
 * Whether autostart, fullscreen and ducking mean anything here. Not
 * `useIsNarrow()`: a desktop window dragged narrow is still a desktop.
 */
export function useIsDesktopHost(): boolean {
  return useHost()?.desktop ?? false;
}

File: src-tauri/src/features/host.rs
//! What the machine underneath actually is.
//!
//! Window width is a different question and lives in `src/layouts`: a desktop
//! window dragged narrow is still a desktop.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub(crate) struct Host {
    /// `linux`, `macos`, `windows`, `android` or `ios`.
    pub(crate) os: &'static str,

    /// Whether autostart, fullscreen and ducking mean anything here.
    pub(crate) desktop: bool,
}

#[tauri::command]
pub(crate) fn host_info() -> Host {
    Host {
        os: std::env::consts::OS,
        desktop: cfg!(desktop),
    }
}
